### ISSUE TITLE
Query result not propagated to component (stuck in loading state).

### ISSUE THREAD

**Issue by Panoplos** [apollographql/react-apollo#1281]
**Intended outcome:**
I have several graphql HOC's, some that have dependencies on the results of others. In one particular case, the query is stuck in a `loading` state, even though the query has completed and the store updated.

Here is the query that is getting stuck.

```es6
const withServiceProvider = compose(
  graphql(ServiceProviderQuery,
    {
      name: 'serviceProvider',
      skip: ({account}) => !account || !account.serviceProvider,
      options: ({account}) => ({
        variables: {
          id: account.serviceProvider && account.serviceProvider.id
        }
      }),
      props: ({ serviceProvider }) => {
        const props = {
          serviceProvider: {
            ...flattenGraphQLProps('serviceProvider', serviceProvider)
          }
        }
        const sp = path(['serviceProvider'], serviceProvider)
        if (!sp) return props
        return {
          serviceProvider: {
            ...props.serviceProvider,
            viewId: spViewId(sp)
          }
        }
      }
    }
  ),
  setPropTypes({
    serviceProvider: PropTypes.object
  })
)
```

Here is the Component that is using it:

```es6
const ServiceProviderContainer = (props: Object): React$Component => {
  const { l10n, serviceProvider } = props
  return (
    <div id='service-provider-container'>
      <Header />
      {
        serviceProvider.loading ?
          <Loader />
        :
          serviceProvider.error ?
            <NotFound message={l10n.localise('sp-not-found', 'This service provider does not exist.')} />
          :
            <Route path='/service-provider/:id/:module?'>
              <ServiceProvider {...props} />
            </Route>
      }
    </div>
  )
}
```

I am composing the component with `withServiceProvider` in the same way I am doing with all my other components that use HOCs.

```es6
export default compose(
  withl10n('service-provider', resourceIds),
  withAccount,
  withServiceProvider,
  ...
)(ServiceProviderContainer)
```

Now, all is great...the loader shows, the query is executed and the results are serialized into the store, 

![redux store updated](https://user-images.githubusercontent.com/2549985/31990047-eebe065a-b9ae-11e7-9142-93cfe832d42f.png)

![apollo store updated](https://user-images.githubusercontent.com/2549985/31990229-736d0c3e-b9af-11e7-8424-093f07342f5f.png)

and normally the component will receive the result. However...

**Actual outcome:**

As you can see below, the component is not getting the final result, and `loading` is never cleared.

![react component not updated](https://user-images.githubusercontent.com/2549985/31990492-38811b8c-b9b0-11e7-89a0-b29a0d2b98f6.png)

**Notes**
1. I have confirmed that `props` is never called with a completed `serviceProvider` (i.e. `data`). It is called twice, each time with the same, as follows:

![prop called](https://user-images.githubusercontent.com/2549985/31991123-f17d9a74-b9b1-11e7-8a4b-4ad01398d270.png)

2. I use this method with many other components, with no issues, so it is really unclear why this one is failing. Any insight into how to debug this would be greatly appreciated.

**Version**
- apollo-client@1.9.3
- react-apollo@1.4.16

**Comment by makamekm** [comment]
The same with:
"apollo-client": "2.0.1"
"react-apollo": "2.0.0"

**Comment by tpinne** [comment]
Having exactly the same problem. But we have one component which behaves correctly even though it is programmed in the same way than the others.

A strange effect to add: The problem only occurs if I navigate between different pages which issue different GraphQL Queries.

No Problem when: 
GQLPage -> NonGQLPage -> GQLPage -> NonGQLPage -> GQLPage ...

Problem when: 
GQLPage -> OtherGQLPage -> GQLPage -> OtherGQLPage -> GQLPage ...
The first call to each page behaves correctly. All subsequent calls are stuck in loading state. FetchPolicy is set to network-only everywhere.

**Comment by bckellerman-soltech** [comment]
I am also having only queries on a single page of my application stuck in loading like @tpinne. The initial query works every time, but on navigation to a new route with a graphql query, and then returning to the previous page with the offending components, those components are stuck in a loading state. I do see the subsequent queries being fired in the network tab, and valid results coming back, but the components are never receiving that data.

On the page this issue is happening on though, there is one query that is working and the others aren't. The only difference between the two is that the working query takes no variables.

This is happening on:
"react-apollo": "1.4.16"
"apollo-client": "1.9.3"

**Comment by bckellerman-soltech** [comment]
After searching more this weekend I found this issue in [apollo-client](https://github.com/apollographql/apollo-client/issues/1186). It looks like the issue is fixed for the client in 2.0, but until doing an upgrade the only fix/hack around this is to put `{ options: { notifyOnNetworkStatusChange: true } }` on all queries having this issue.

**Comment by Panoplos** [comment]
Nope, still happening for me on 2.0...

**Comment by ajdegol** [comment]
Am getting this problem on version 2.0 also.

There seems to be two conditions for it:
1. Pushing to the (react router) route where the component is.
2. Cache has to be cleared beforehand.

Happy to provide as many more details as you need.

**Comment by Panoplos** [comment]
Ok. Confirmed this is fixed with 2.0.1.
